src/control/server/instance.go

```diff
@@ -218,9 +218,47 @@ func (ei *EngineInstance) determineRank(ctx context.Context, ready *srvpb.Notify
 	return r, resp.LocalJoin, nil
 }
 
+func (ei *EngineInstance) isFaultDomainChanging(sb *Superblock) bool {
+	if ei.hostFaultDomain == nil && sb.HostFaultDomain == "" {
+		return false
+	}
+
+	return !(ei.hostFaultDomain.String() == sb.HostFaultDomain)
+}
+
+func (ei *EngineInstance) updateFaultDomainInSuperblock() error {
+	superblock := ei.getSuperblock()
+	if superblock == nil {
+		return errors.New("nil superblock while updating fault domain")
+	}
+
+	if !ei.isFaultDomainChanging(superblock) {
+		return nil
+	}
+
+	ei.log.Infof("instance %d setting host fault domain to %q (previously %q)",
+		ei.Index(), ei.hostFaultDomain, superblock.HostFaultDomain)
+
+	newDomainStr := ""
+	if ei.hostFaultDomain != nil {
+		newDomainStr = ei.hostFaultDomain.String()
+	}
+	superblock.HostFaultDomain = newDomainStr
+
+	ei.setSuperblock(superblock)
+	if err := ei.WriteSuperblock(); err != nil {
+		return err
+	}
+	return nil
+}
+
 // handleReady determines the instance rank and sends a SetRank dRPC request
 // to the Engine.
 func (ei *EngineInstance) handleReady(ctx context.Context, ready *srvpb.NotifyReadyReq) error {
+	if err := ei.updateFaultDomainInSuperblock(); err != nil {
+		ei.log.Error(err.Error()) // nonfatal
```

Comment:
Is this really a nonfatal error? Seems like failing to persist important metadata is something we wouldn't want to gloss over. Might want to rework the logic so that `updateFaultDomainInSuperblock()` is only called when it's expected to succeed and then this error is more meaningful.

Reply:
I just thought this metadata in particular isn't that important. The only reason to track it in the superblock is to know that it changed. Practically speaking, if there's a problem writing the superblock, it will be caught while persisting the rank (which *is* important).